### ISSUE TITLE
Bad inference when using void and &&

### ISSUE THREAD

**Issue by AlexGalays** [Microsoft/TypeScript#8867]
```javascript
const a: void = undefined;
const b: number = 33;

const z = a && b;
```

z is always inferred to be of the same type as b, even though a is undefined.

**Comment by Arnavion** [comment]
Without `--strictNullChecks` there is nothing wrong. `&&` is allowed when the left expression is of type `void` to maintain compatibility with JS, and `undefined` is a valid value for `number`.

With `--strictNullChecks` it should probably be inferred as `number | undefined | null`, but it's still inferred as `number`. That looks like a bug.

**Comment by Arnavion** [comment]
For future reference, the bug I noted above doesn't happen if `a` is of type `undefined` instead of `void`. I imagine it's probably best to never use `void` as a holder for `undefined` values to reap the benefits of `--strictNullChecks`

**Comment by AlexGalays** [comment]
Not sure what you mean, undefined is not a type?  Are you saying void should never be used at all in type annotations and we should just rely on type inference?

Anyway, I'm not sure why this got closed without much explanation :(

**Comment by kitsonk** [comment]
I think @Arnavion assumed TypeScript@next usage (as you didn't specify a version of TypeScript in the issue, though the issue template requests one).  In TypeScript 2.0, `undefined` and `null` are types that are removed from all other types (except `any`) and can be dealt with better.

So, if you feel like you need to be neurotic in your type annotations:

```typescript
const a: undefined = undefined;
```

Is a better way of representing `undefined`, but of course it being a `const`, therefore disallowing reassignment, the type annotation it entirely superfluous.

TypeScript 2.0 introduces the compiler flag `--strictNullChecks` and @Arnavion noted that with `--strictNullChecks` the `void` type appears to not be handled properly, which he states is likely a bug, though I think @mhegazy might have not have noticed that, trying to quickly triage the name issues that are opened here.  On the other hand, the new types of `undefined` and `null` are handled properly, so it might not be a bug.

In 1.8, there is no `undefined` type, but the compiler is still working correctly, since `number` could be expressed as `number | undefined | null` in TypeScript 2.0 terms as all types are inclusive of those values, because JavaScript is a crazy language.  There is no concept of a hard "bottom" type in TypeScript 1.8.  `void` is only used to imply that no value is likely to be returned from an operation, but upon assignment, it doesn't "hold".

So in summary, there is an improvement of this in TypeScript 2.0 that lets you better model this, but technically TypeScript 1.8 isn't broken either.

**Comment by AlexGalays** [comment]
I'm now happily using TS 2.0 and use the undefined type whenever it applies, however I still struggle to see what's the difference between undefined and void. Is that documented somewhere ?

**Comment by kitsonk** [comment]
Like Cher, I suspect that if the TypeScript team could rollback time, `void` would not exist and `undefined` and `null` would have been distinct types from the start.

`void` is intended to denote a return from a function where from an interface perspective, no one should depend upon any return.  It wasn't really a type, in the way `undefined` is.  Because of the significant breaking change of `strictNullChecks` this was introduced under a flag to avoid breaking all the stuff people already had (and make transition from JavaScript to TypeScript easy and potentially incremental).

I am not sure it documented anywhere specifically yet.
